pytools/graph.py

```diff
@@ -566,24 +567,39 @@ def undirected_graph_from_edges(
 
 def get_reachable_nodes(
         undirected_graph: GraphT[NodeT],
-        source_node: NodeT) -> frozenset[NodeT]:
+        source_node: NodeT,
+        exclude_nodes: Optional[set[NodeT]] = None) -> frozenset[NodeT]:
     """
     Returns a :class:`frozenset` of all nodes in *undirected_graph* that are
-    reachable from *source_node*.
+    reachable from *source_node*. Reachability of nodes from *source_node* can
+    be restricted by supplying a set of nodes to be excluded from consideration.
+    Effectively, this amounts to traversing partitions of *undirected_graph*
+    with partition points defined by *exclude_nodes*.
     """
     nodes_visited: set[NodeT] = set()
+    reachable_nodes: set[NodeT] = set()
     nodes_to_visit = {source_node}
 
+    if exclude_nodes is None:
+        exclude_nodes = set()
+
     while nodes_to_visit:
         current_node = nodes_to_visit.pop()
         nodes_visited.add(current_node)
 
-        neighbors = undirected_graph[current_node]
-        nodes_to_visit.update({node
-                               for node in neighbors
-                               if node not in nodes_visited})
+        if current_node in exclude_nodes:
+            continue
```

Comment:
This seems redundant with the check below, except for the case of the initial node, which can be handled above.